core/src/main/java/com/linecorp/armeria/server/ServiceBindingBuilder.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.nio.charset.Charset;
+import java.time.Duration;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.logging.ContentPreviewerFactory;
+
+/**
+ * A builder class for binding a {@link Service} fluently. This class can be instantiated through
+ * {@link ServerBuilder#route()}. You can also configure a {@link Service} using
+ * {@link ServerBuilder#withRoute(Consumer)}.
+ *
+ * <p>Call {@link #build(Service)} to build the {@link Service} and return to the {@link ServerBuilder}.
+ *
+ * <pre>{@code
+ * ServerBuilder sb = new ServerBuilder();
+ *
+ * sb.route()                                      // Configure the first service.
+ *   .get("/foo/bar")
+ *   .consumes(JSON, PLAIN_TEXT_UTF_8)
+ *   .produces(JSON_UTF_8)
+ *   .requestTimeoutMillis(5000)
+ *   .maxRequestLength(8192)
+ *   .verboseResponses(true)
+ *   .contentPreview(500)
+ *   .build((ctx, req) -> HttpResponse.of(OK));    // Return to the ServerBuilder.
+ *
+ * // Configure the second service with Consumer.
+ * sb.withRoute(builder -> builder.path("/baz")
+ *                                .methods(HttpMethod.GET, HttpMethod.POST)
+ *                                .build((ctx, req) -> HttpResponse.of(OK)));
+ * }</pre>
+ *
+ * @see VirtualHostServiceBindingBuilder
+ */
+public final class ServiceBindingBuilder extends AbstractServiceBindingBuilder {
+
+    private final ServerBuilder serverBuilder;
+
+    ServiceBindingBuilder(ServerBuilder serverBuilder) {
+        this.serverBuilder = requireNonNull(serverBuilder, "serverBuilder");
+    }
+
+    @Override
+    public ServiceBindingBuilder path(String pathPattern) {
+        return (ServiceBindingBuilder) super.path(pathPattern);
```

Comment:
Can we add a template parameter to the abstract class instead of having to add these in children?

Reply:
I couldn't use it because of the Javadoc API reference. 😰 
The abstract parent class is package-private so this child inherits the Javadoc from the parent. In the rendered API reference page, the return type is the template parameter (e.g `T` not `ServiceBindingBuilder`).
And I couldn't add a description like `@param <T> the self type` to the child  Javadoc.
We consider using self-describing template parameter such as `SELF` instead of `T` like AssertJ does. But probably it's not a common, so we chose this painful way. 😢